### PR TITLE
Fixing "Ambiguous mapping methods found for mapping property" error w…

### LINKED ISSUE #5403: Ambiguous mapping methods found for mapping property while compiling the mappers of DTOs created using mapstruct
##### **Overview of the issue**

"Ambiguous mapping methods found for mapping property" while compiling the mappers of DTOs created using mapstruct. This is happening when we have mix of MTM and OTM relationships on the same object.

##### **Motivation for or Use Case**

This causes the gradlew(or mvnw) fail to start with compilation errors.

##### **Reproduce the error**

Bellow is the JDL file

```
entity WishList {
    name String,
    description String
}

relationship OneToMany {
    Profile {
        wishList
    }
    to WishList
}

relationship ManyToMany {
    WishList {
        listing
    }
    to Listing
}


entity Profile {
    name String,
    description String
}

relationship OneToOne {
    Profile {
        user
    }
    to User
}

relationship ManyToOne {
    Listing {
        host(name)
    }
    to Profile
}


entity Listing {
    name String,
    description String
}



dto * with mapstruct

// Set service options to all except few
service all with serviceImpl 
// Set an angular suffix
angularSuffix * with mySuffix`
```

compilation fails for WishListMapper with Ambiguous profileFromId method.

##### **Related issues**
Yes. #2006 and #37 on hipster-core

##### **Suggest a Fix**

Previous comments seems to attribute this issue to mapstruct but i believe the issue seems to be cause of the way we are generating the mappers fromId.

we can fix the issue by refactoring the way we are generating the fromId.

##### **JHipster Version(s)**

4.x

##### **JHipster configuration**
```
{
  "generator-jhipster": {
    "jhipsterVersion": "4.0.8",
    "baseName": "fail",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "gradle",
    "enableSocialSignIn": true,
    "clientFramework": "angular2",
    "useSass": false,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": false
  }
}
```
##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

```json
{
    "fluentMethods": true,
    "relationships": [
        {
            "relationshipType": "many-to-many",
            "otherEntityRelationshipName": "",
            "relationshipName": "listing",
            "otherEntityName": "listing",
            "otherEntityField": "id",
            "ownerSide": true
        },
        {
            "relationshipName": "profile",
            "otherEntityName": "profile",
            "relationshipType": "many-to-one",
            "otherEntityField": "id"
        }
    ],
    "fields": [
        {
            "fieldName": "name",
            "fieldType": "String"
        },
        {
            "fieldName": "description",
            "fieldType": "String"
        }
    ],
    "changelogDate": "20170311065829",
    "entityTableName": "wish_list",
    "dto": "mapstruct",
    "pagination": "no",
    "service": "serviceImpl",
    "angularJSSuffix": "mySuffix"
}
```

```json
{
    "fluentMethods": true,
    "relationships": [
        {
            "relationshipType": "one-to-one",
            "relationshipName": "user",
            "otherEntityName": "user",
            "otherEntityField": "id",
            "ownerSide": true,
            "otherEntityRelationshipName": "profile"
        },
        {
            "relationshipType": "one-to-many",
            "relationshipName": "wishList",
            "otherEntityName": "wishList",
            "otherEntityRelationshipName": "profile"
        }
    ],
    "fields": [
        {
            "fieldName": "name",
            "fieldType": "String"
        },
        {
            "fieldName": "description",
            "fieldType": "String"
        }
    ],
    "changelogDate": "20170311065829",
    "entityTableName": "profile",
    "dto": "mapstruct",
    "pagination": "no",
    "service": "serviceImpl",
    "angularJSSuffix": "mySuffix"
}
```
```json
{
    "fluentMethods": true,
    "relationships": [
        {
            "relationshipType": "many-to-one",
            "relationshipName": "host",
            "otherEntityName": "profile",
            "otherEntityField": "name"
        }
    ],
    "fields": [
        {
            "fieldName": "name",
            "fieldType": "String"
        },
        {
            "fieldName": "description",
            "fieldType": "String"
        }
    ],
    "changelogDate": "20170311065829",
    "entityTableName": "listing",
    "dto": "mapstruct",
    "pagination": "no",
    "service": "serviceImpl",
    "angularJSSuffix": "mySuffix"
}
```

##### **Browsers and Operating System**

Any Os and Browser

- [X] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by sreekrishna**
…hile compiling the mappers of DTOs created using mapstruct

Changed EntityMapper generation to create the source mapper than creating in target mapper class and
 add the appropriate mappers in the maptstruct @Uses.This will eliminate the ambiguous methods in target mapper class

Fix #5403

### CODE REVIEW

File: generators/entity/templates/server/src/main/java/package/service/mapper/_EntityMapper.java [hunk 1]
```diff
@@ -47,9 +47,10 @@
     List<<%= entityClass %>> <%= entityInstance %>DTOsTo<%= entityClassPlural %>(List<<%= entityClass %>DTO> <%= entityInstance %>DTOs);
 
     /**
-     * generating the fromId for all mappers, as the class has relationship to it might need it, instead of
+     * generating the fromId for all mappers if the databaseType is sql, as the class has relationship to it might need it, instead of
      * creating a new attribute to know if the enitity has any relationship from someother entity
```
Comment (ruddell) [inline generators/entity/templates/server/src/main/java/package/service/mapper/_EntityMapper.java]: Spelling `enitity` -> `entity` and `someother` -> `some other`

I'll test this today

**Comment by ruddell** [comment]
Tested and it works.  This solves a long-time issue with DTOs, and there are tests!  Thanks for the PR.

Could you please shorten the comment to something like below (the sql note isn't important because it only shows up in sql apps):
`this method is used by related entity mappers`

**Comment by jdubois** [comment]
Excellent!!!!
I'm merging this as @ruddell tested it, this is really a great PR: that part of the code is hard, and you even did tests, that's awesome @sreekrishna !

**Comment by sreekrishna** [comment]
glad that i was able to help. thanks @jdubois and @ruddell

**Pull request merged**